### ISSUE TITLE
High (too high?) memory usage when running the full tournament

### ISSUE THREAD

**Issue by drvinceknight** [Axelrod-Python/Axelrod#520]
When I update the tournament results I get a crash:

```
tournament(gh-pages) ✗: ./update_results.sh 0 20
-- Regenerating noisy tournament figures
/usr/local/lib/python3.4/dist-packages/matplotlib/backends/backend_gtk3agg.py:18: UserWarning: The Gtk3Agg backend is known to not work on Python 3.x with pycairo. Try installing cairocffi.
  "The Gtk3Agg backend is known to not work on Python 3.x with pycairo. "
Starting basic_strategies tournament: Turns: 200, Repetitions: 20, Strategies: 8.
Finished all basic_strategies tasks in 8.4s
Starting ordinary_strategies tournament: Turns: 200, Repetitions: 20, Strategies: 113.
Finished all ordinary_strategies tasks in 1370.4s
Starting strategies tournament: Turns: 200, Repetitions: 20, Strategies: 123.
Traceback (most recent call last):
  File "run_axelrod", line 132, in <module>
    run_tournaments(**kwargs)
  File "/home/vince/src/Axelrod/axelrod/utils.py", line 82, in run_tournaments
    manager.run_tournaments()
  File "/home/vince/src/Axelrod/axelrod/tournament_manager.py", line 65, in run_tournaments
    self._run_single_tournament(tournament)
  File "/home/vince/src/Axelrod/axelrod/tournament_manager.py", line 96, in _run_single_tournament
    self._generate_output_files(tournament, ecosystem)
  File "/home/vince/src/Axelrod/axelrod/tournament_manager.py", line 121, in _generate_output_files
    image_format=self._image_format)
  File "/home/vince/src/Axelrod/axelrod/tournament_manager.py", line 143, in _save_plots
    self._save_plot(figure, file_name)
  File "/home/vince/src/Axelrod/axelrod/tournament_manager.py", line 164, in _save_plot
    figure.savefig(file_name, bbox_inches='tight', dpi=dpi)
  File "/usr/local/lib/python3.4/dist-packages/matplotlib/figure.py", line 1565, in savefig
    self.canvas.print_figure(*args, **kwargs)
  File "/usr/local/lib/python3.4/dist-packages/matplotlib/backend_bases.py", line 2180, in print_figure
    **kwargs)
  File "/usr/local/lib/python3.4/dist-packages/matplotlib/backends/backend_svg.py", line 1206, in print_svg
    return self._print_svg(filename, svgwriter, fh_to_close, **kwargs)
  File "/usr/local/lib/python3.4/dist-packages/matplotlib/backends/backend_svg.py", line 1234, in _print_svg
    self.figure.draw(renderer)
  File "/usr/local/lib/python3.4/dist-packages/matplotlib/artist.py", line 61, in draw_wrapper
    draw(artist, renderer, *args, **kwargs)
  File "/usr/local/lib/python3.4/dist-packages/matplotlib/figure.py", line 1159, in draw
    func(*args)
  File "/usr/local/lib/python3.4/dist-packages/matplotlib/artist.py", line 61, in draw_wrapper
    draw(artist, renderer, *args, **kwargs)
  File "/usr/local/lib/python3.4/dist-packages/matplotlib/axes/_base.py", line 2324, in draw
    a.draw(renderer)
  File "/usr/local/lib/python3.4/dist-packages/matplotlib/artist.py", line 61, in draw_wrapper
    draw(artist, renderer, *args, **kwargs)
  File "/usr/local/lib/python3.4/dist-packages/matplotlib/image.py", line 394, in draw
    renderer.draw_image(gc, l, b, im)
  File "/usr/local/lib/python3.4/dist-packages/matplotlib/backends/backend_svg.py", line 837, in draw_image
    _png.write_png(np.array(im)[::-1], bytesio)
MemoryError
```

This is running on ubuntu 14.04 with 16GB of Ram (the same machine I've used previously). You'll see there that I'm only actually running 20 repetitions (I had similar crashes for 100 and 50 reps).

I'm about to open a PR that makes improvements anyway to memory (the crash happens for me on `master` as well as on that branch although that branch shows a much lower memory consumption) but it would be good to check if this is just my machine: @marcharper could you run the results again (first on `master`?).

**Comment by marcharper** [comment]
Will do. We should consider running a full tournament as part of the tests. I know that's a drag because it takes a long time, but we could change the travis script to fail on the first failure and run the full tournament test last.

**Comment by drvinceknight** [comment]
We already do, it's in the integration tests. What we don't do is run it for the full rep and turn count which I don't think we should (it would timeout on travis). The issue here is simply the number of reps... :) 

Standard full tournament seems to have run fine (from master) on our university cluster. Will also set it up on the 520 branch which I think is an improvement anyway...

**Comment by marcharper** [comment]
It's taking a lot of memory -- more than 1gb per thread. I would guess this is due to all the object creation and the returning of the full interaction history. It's not killing my system (also 16gb) but it could get close if I have a lot of other stuff running.

**Comment by drvinceknight** [comment]
Yeah... If and when you get a chance if you could let me know if 520 is any
better...

It crashed on me with my 16GB doing nothing else..

I'm going to work on this more today... It would be easy enough to write
the interaction history to file and then analysis can be done at a second
stage... I might try that...

It's all humming along on the cluster and getting results out...

**Comment by drvinceknight** [comment]
Closed by 4f90b42

**Comment by marcharper** [comment]
I'm getting a lot of memory usage with a tournament with several lookerups (129 strategies) of 100 repetitions and 200 turns:
- 7 Gb on the main thread
- 1 to 2 Gb on the aux threads

I have 16 Gb of RAM and I'm starting to hit the swap :/

**Comment by drvinceknight** [comment]
This doesn't surprise me. On https://github.com/Axelrod-Python/Axelrod/pull/523 I wrote the memory usage for each tournament (reported from running on the cluster). Writing to file and not creating  the results directly might help but still with 100 repetitions the usage is just very high. You could do two separate runs of 50 reps and combine the data?

**Comment by marcharper** [comment]
I can do that.

**Comment by drvinceknight** [comment]
If we want the tournament class to keep the interactions attribute (which we do I believe), I'm not sure there's much we can do... (It's just big...?). #523 was about making it so we didn't _have to_ rely on memory completely... 

Not sure if there's anything else we can do? (but other eyes on it would be worthwhile)

**Comment by drvinceknight** [comment]
I have an idea! 

We transpose the format of the data file. Currently rows are matches and columns are repetitions. Swap that so it becomes rows for repetitions and columns for matches (this would also have the advantage of making it easier to combine tournament runs which is in effect what repetitions are). 

Then, the minor change to the tournament would be that if writing to file is triggered with `play` then we don't append the interactions, we just write them to file as we go. Thus we're not writing to file at the end but throughout which should keep memory down and basically make the hunger of repetitions irrelevant.

**Comment by drvinceknight** [comment]
That's really how I should have done it in the first place... :/

**Comment by drvinceknight** [comment]
@marcharper could you try what I've got on #560 (wip) it improves how tournaments are written to file (not once at the end but as we go). This could theoretically (hopefully) mean that if writing to file, the number of repetitions of a tournament will have a very low impact on memory...

**Comment by marcharper** [comment]
This seems to help a bit but it's still taking over 1 Gb per thread.

I've run into situations like this before, and I think a good approach is to use a thread pool -- I think the stated reason for not using a pool can be handled now (from tournament.py):
```
        # At first sight, it might seem simpler to use the multiprocessing Pool
        # Class rather than Processes and Queues. However, Pool can only accept
        # target functions which can be pickled and instance methods cannot.
```
Either we can pickle them now or we can write a module-level "play match" function, which is probably not much more complicated than this:
```
def play_match(match, queue):
     results = match.play()
     queue.put((match_info, results))

````
Then we have the calling thread write to disk constantly by consuming the queue and sleeping for ~1 sec if the queue is empty. This should prevent the threads from ever taking up too much memory -- the limiting factor should be the total history of a single match.

**Comment by drvinceknight** [comment]
Sounds good to me. Would you be able to give that a go? I know you're busy
so I can if not but it probably wouldn't be for a week or so (no huge
rush).

**Comment by drvinceknight** [comment]
Side thought: is 1GB per thread actually that bad? Given the combinatorial number of matches it doesn't seem that high to me. 

Obviously, if we can reduce sounds good. Rereading,  is your idea to not parallelize at the repetitions level but at the match level? (Just checking Iunderstand.) :)

**Comment by marcharper** [comment]
>  is 1GB per thread actually that bad?
I stopped it when it hit about 1.5 GB per thread and was continuing to grow. It simply grows less quickly now.
> Rereading, is your idea to not parallelize at the repetitions level but at the match level?
Yes, IMO the match is the central object now, and it's the smallest unit that makes sense to parallelize.

**Comment by drvinceknight** [comment]
> I stopped it when it hit about 1.5 GB per thread and was continuing to grow. It simply grows less quickly now.

Hmmm, that's a shame, I was hoping it wouldn't grow... Oh well.

> Yes, IMO the match is the central object now, and it's the smallest unit that makes sense to parallelize.

Cool. That's nice I guess because we could then still parallelise tournaments with 1 repetition but a million strategies (theoretically). As the Match itself is quite 'short', is there a risk of diminishing returns (I'm on board, just wondering). 

I guess we would still write to file at the 'repetition' level? Like I've done in #560 or are you thinking of writing at the Match level too?

Theoretically the data file format could be even simpler:

```
Index Pair, Name Pairs, CCDDCCDDCDC
...
```

So just the 3 'columns', where for repetitions there would be rows that just 'repeat' a given number of times... That again is a nice move away from any assumption about the structure...

Similar question to above though: would writing after each Match (which would result in many many writes) be costly?

**Comment by marcharper** [comment]
You wouldn't write after each match, you would have another thread that does all the writing to disk, and it would be buffered automatically by python. Sometimes this would be many matches, sometimes few -- it would depend on the match time. I don't think that would be the bottleneck.

If you are worried about diminishing returns, we could play all `repetitions` of the matches in the same thread, i.e. all 100 matches between A and B are played one after the other by the same worker:

```
def play_match(match, queue, repetitions=1):
     for repetition in repetitions:
         results = match.play()
         queue.put((match_info, results))
```

I think the simpler format is better. I can pick this one up.

**Comment by drvinceknight** [comment]
> You wouldn't write after each match, you would have another thread that does all the writing to disk, and it would be buffered automatically by python. Sometimes this would be many matches, sometimes few -- it would depend on the match time. I don't think that would be the bottleneck.

Cool.

> 
> If you are worried about diminishing returns, 

Just thinking of overheads but perhaps that's not a problem. Whatever you feel is best :)

> we could play all repetitions of the matches in the same thread, i.e. all 100 matches between A and B are played one after the other by the same worker:
> 
> def play_match(match, queue, repetitions=1):
>      for repetition in range(repetitions):
>          results = match.play()
>          queue.put(results)

Ooo I like that (I forget if we'd need a reset method call in there somewhere?). At the moment we're calling the `match_generator` to regenerate matches for each repetition . Actually just repeating the plays of the match and extracting the information seems better.

> I think the simpler format is better. 

I agree. So like currently here: http://axelrod.readthedocs.org/en/latest/tutorials/advanced/reading_and_writing_interactions.html but without the extra columns for each repetition (just a repetition of the entire row).

> I can pick this one up.

Sweet. I'll go ahead and close #560, I think what you're going to do pretty much supersedes that. Probably worth holding fire on #558 too (will remove the ready for review tag) as that will no doubt need to change after this.

**Comment by marcharper** [comment]
Here's what I'm proposing. Just a few notes:
* This uses less than 100MB per thread (on my machine) for 1000 repetitions
* I'm using `json` for output temporarily because I haven't messed with the output format
* All this really does is keep a pool of processes populated and writes the contents of a queue to a file
when not spawning new threads

```
import json
import time

from multiprocess import Process, Queue

import axelrod as axl

def match_generator(players, turns=100):
    for player1 in players:
        for player2 in players:
            players = (player1.clone(), player2.clone())
            match = axl.Match(players, turns)
            yield match

# Could probably just use `Process(target=func, args=(match, queue, repetitions))`
class MatchProcess(Process):
    def __init__(self, match, queue, repetitions=1000):
        Process.__init__(self)
        self.match = match
        self.queue = queue
        self.repetitions = repetitions

    def run(self):
        for repetition in range(self.repetitions):
            results = self.match.play()
            self.queue.put(results)

class ProcessManager(object):
    def __init__(self, players, max_processes=4, filename="results.out"):
        self.processes = []
        self.max_processes = max_processes
        self.queue = Queue()
        self.file = open("filename", 'w')
        self.matches = list(match_generator(players))

    def clean_pool(self):
        terminated = []
        for p in self.processes:
            # Look for naturally terminated processes.
            if not p.is_alive():
                terminated.append(p)
        for t in terminated:
            self.processes.remove(t)

    def run(self):
        # Add workers
        while len(self.matches) > 0:
            while len(self.processes) < self.max_processes:
                p = MatchProcess(self.matches.pop(), self.queue)
                self.processes.append(p)
                p.start()
            # Clean Pool
            self.clean_pool()
            # Write Queue to disk
            # Could have another process just for this
            while self.queue.qsize() > 0:
                results = self.queue.get()
                self.file.write(json.dumps(results))
        time.sleep(0.2)

if __name__ == "__main__":
    players = [s() for s in axl.ordinary_strategies]
    p = ProcessManager(players)
    p.run()
```

**Comment by drvinceknight** [comment]
> This uses less than 100MB per thread (on my machine) for 1000 repetitions

Hahaha :bow: :clap: :+1: 

Cool.

Would the idea be that some/all of that be in new module (`parallel_matches.py` or whatever?) and called up by the `Tournament` class? So the `Tournament` class would see less of the complicated stuff about the parallel processes?

Perhaps (I know this isn't meant to be final) the `match_generator` could be called something else so that it does not lead to confusion with the `MatchGenerator` class? Perhaps `match_repeater`?

Re `json`, just a thought (I know you're just using that temporarily). It's a format I haven't used much but my very basic understanding is that it's a web friendlier format, would that perhaps make more sense as a format for the tournaments? Is it friendly enough to the wider research community would be my main concern, perhaps csv is best to stay with.. Just a thought...

If you haven't already, could you check this works well with the decorators?

**Comment by drvinceknight** [comment]
> Perhaps (I know this isn't meant to be final) the match_generator could be called something else so that it does not lead to confusion with the MatchGenerator class? Perhaps match_repeater?

Looking again, I'm not entirely sure how this sits with the current `MatchGenerator` (which allows us to create structures where everyone would not play everyone). Perhaps you just had that temporarily there and are planning on using the original `MatchGenerator`? Basically I think that question by me is to be ignored :)

**Comment by marcharper** [comment]
* It's not final
* We can easily modify so that the `ProcessManager` can take any generator, that's no problem, I just made a quick one for demo purposes
* I'm thinking that this code would go into `match.py` wrapped in a function called something like `play_matches_parallel` that takes a match generator/iterator
* I think Tournament would call this function with its match generator
* I just used json for demo purposes out of lazyness because python complained when I just tried to write the match output directly to a file, and this was the easiest way to serialize the output to a string.

**Comment by drvinceknight** [comment]
Cool! All sounds great! Looking forward to it :)

**Comment by drvinceknight** [comment]
> Actually just repeating the plays of the match and extracting the information seems better.

Just a thought (you might have already got this covered): the prob end matches are matches of a given length: currently the generator re samples the length each time. So just repeating a play of a given match would have them be all the same length. You might have already thought about this... (Perhaps modifying the structure so having the generator yield matches for a given tuple of players would be a way to do this?)

Slightly related (feel free to ignore this, was just crossing my mind earlier). The interactions that gets passed to the results (and what the reading from data file creates): currently a list of dictionaries mapping tuples to outcomes of matches:

```
[{(0,1), [('C', 'D'), ('D','C')],....},
 {(0,1), [('C', 'D'), ('D','C')],....}]
```

Could perhaps be nicer to just be a dictionary mapping tuples to lists of ouctomes of matches... This would actually make the analysis a tiny bit simpler...

```
{(0,1): [[('C', 'D'), ('D','C')], [('C', 'D'), ('D','C')]], ....}
```

It would not take much to modify the results set to cope with this.

**Comment by drvinceknight** [comment]
(Both of those are just ideas: feel free to throw them away :))

**Comment by marcharper** [comment]
I liked the earlier suggestion of a simple csv layout:
```
Player Index,Player Name,Opponent Index,Opponent Name,Player Actions,Opponent Actions
1,2,TFT,Defector,CDDD,DDDD
```
It's simple to go back and forth from this format to the one described above, and this format is better for data sharing purposes IMO, but I'm open to other considerations. One issue is that the matches are not written to disk in sorted order.

I think the way to go for the probabilistically ending matches is to use a modified generator.

**Comment by drvinceknight** [comment]
I completely agree about the csv. :+1:

My suggestion was for the result set input/interactions object that it uses. I gave an example of the status quo (list of dicts)  and what it could be (dict with lists as values, basically more along the lines of not repeating a tournament but repeating a match). Don't worry though if I'm not being clear, not important at this stage :) That could be a simplifying refactor of the results set at a later pointpoint.

**Comment by drvinceknight** [comment]
Just back from a good 🐕  walk.

So I don't think the following is actually a good idea (I know I said I originally liked this):

```
    def run(self):
        for repetition in range(self.repetitions):
            results = self.match.play()
            self.queue.put(results)
```

This goes back to the many conversations we've had about the potential for weird and wonderful formats of tournaments.

My suggestion (to follow), **is is all based on the assumption that you can write some magic that parallelises at the Match level**, I might well be misunderstanding things about that so bare with me :) 

Currently, the match generator (and a lot of other things) is based on the tournament being the 'fundamental unit': it yields, tuples of players and matches. This represents a tournament and (currently) the generator is repeatedly called.

I think we should modify the generator to incorporate the repetitions:

```
    def build_matches(self, noise=0, repetitions=1): # <- adding an argument
        for rep in range(repetitions):  # <- adding a loop (whether or not it's at this level or nested below)
            for player1_index in range(len(self.players)):
                for player2_index in range(player1_index, len(self.players)):
                    pair = (
                        self.players[player1_index], self.opponents[player2_index])
                     match = self.build_single_match(pair, noise)
                     yield (player1_index, player2_index), match
```

This means that the `build_matches` method would be called by the Tournament to create the generator once as it holds all the matches to be played. So theoretically we could imagine [creating new tournaments](http://axelrod.readthedocs.org/en/latest/tutorials/advanced/making_tournaments.html) where the repetitions of matches between two players wasn't even a constant (for some reason?) or whatever.

That suggestion has the advantage of keeping the generator a simple construct for building 'everything we want to see played'.

Then, when it comes to getting the interactions from a match out of the tournament:

1. If we're writing to file we write to file in the format we both like.
2. If we're not writing to file: we replace the current `tournament.interactions` (currently a list of dictionaries, where each dictionary represents a repetition) with a single dictionary that maps tuples of players to lists of outcomes of matches (the list representing the repetitions), then we simply have something like:

```
tuple, interaction = MarcsMagicThing.return_outomes()
if filename is None:
    try:
        self.interactions[tuple].append(interaction)
    except KeyValue:
        self.interactions[tuple] = [interaction]

else:
    self.write_to_file(tuple, interaction)
```

Thus the order of which things are played is not important. We'd have something similar to that in the `ResultsSetFromFile` (so again the order is irrelevant).

This, is in line with the change I suggested above (which is meant to make things easier with the csv file format we both like) in what the `ResultsSet` assumes it's getting. The current dictionaries in the list all have the same keys, the only reason it's done that way is because of the original 'unit building block' being a tournament...

**Comment by drvinceknight** [comment]
Here's a modified version of your code above that I believe would do this: https://gist.github.com/9242ec149030433b800ea1ba71411c89

I wonder if there's a way leaving the generator as is? (Not transforming to a list before passing it to `MarcsMagicThing`).

---

(FYI I'm getting a 

```
    p.run()
  File "520.py", line 57, in run
    while self.queue.qsize() > 0:
  File "/usr/local/lib/python2.7/site-packages/multiprocess/queues.py", line 143, in qsize
    return self._maxsize - self._sem._semlock._get_value()
```
)

**Comment by marcharper** [comment]
I'm with you, we can definitely keep the matches as a proper generator -- again my snippet above was a quick proof of concept adapted from another project and not really well-fleshed out, and it was easier just to make it into a list.

One thing I still need to figure out is how to group the matches together for the probabilistic tournament to take advantage of the cache. I think an optional match length generator could do the trick, and I'm trying to figure out how to handle that case along with the repetitions case nicely... it may be better for the generator to serve up collections of matches or parameters.

I've used generators heavily for this sort of thing several times in the past, just give me a little time to get it sorted and tested.

Here's what I've got at the moment:
```
# Can definitely do better here, with turns a constant or an iterable specifying
# match lengths, and repetitions
def match_generator(players, turns=100):
    for player1 in players:
        for player2 in players:
            players = (player1.clone(), player2.clone())
            match = axl.Match(players, turns)
            yield match

# To be used with Process(target=play_matches, ...)
# Could modify this to take a chunk of matches say 100 at a time to reduce
# overhead
def play_matches(match, queue, repetitions=1):
    for _ in range(repetitions):
        results = match.play()
        queue.put(results)

# This is mostly the same
class ProcessManager(object):
    def __init__(self, matches, repetitions=1, queue=None, max_processes=4,
                 filename="results.out"):
        self.max_processes = max_processes
        self.matches = matches
        self.repetitions = repetitions
        if queue:
            self.queue = queue
        else:
            self.queue = Queue()
        self.file = open("filename", 'w')
        self.processes = []

    def clean_pool(self):
        terminated = []
        for p in self.processes:
            # Look for naturally terminated processes.
            if not p.is_alive():
                terminated.append(p)
        for t in terminated:
            self.processes.remove(t)

    def run(self):
        while len(self.matches) > 0:
            # Add workers
            while len(self.processes) < self.max_processes:
                if len(self.matches) == 0:
                    break
                match = self.matches.pop()
                p = Process(target=play_matches,
                            args=(match, self.queue, self.repetitions))
                self.processes.append(p)
                p.start()
            # Write Queue to disk
            # Could have another process just for this
            while self.queue.qsize() > 0:
                results = self.queue.get()
                self.file.write(json.dumps(results))
            if (len(self.matches) == 0) and (len(self.processes) == 0):
                break
            time.sleep(0.2)
            # Clean Pool
            self.clean_pool()
        return self.queue

# Here we'd pass in a matches generator instead
def play_matches_parallel(matches, repetitions=1, queue=None, filename=None,
                          max_processes=4):
    pm = ProcessManager(matches, repetitions=repetitions, queue=queue,
                        filename=filename, max_processes=max_processes)
    queue = pm.run()
    return queue

if __name__ == "__main__":
    players = [s() for s in axl.ordinary_strategies]
    matches = list(match_generator(players))
    queue = play_matches_parallel(matches, repetitions=100)
```

We can also have a version of ResultsSet that consumes the queue as it runs rather than write everything to disk, but let's get the basics worked out first.

**Comment by drvinceknight** [comment]
Completely realise I'm commenting on temporary stuff (sorry! :)).

All sounds great. 

> I think an optional match length generator could do the trick

There is one in the corresponding `MatchGenerator` if it's helpful. So that should work out of the box with my suggestion above for incorporating the repetitions in the match generator (I see that in your current version you're still looping the plays of the Matches): https://github.com/Axelrod-Python/Axelrod/blob/master/axelrod/match_generator.py#L113

> I've used generators heavily for this sort of thing several times in the past, just give me a little time to get it sorted and tested.

Absolutely! 

Sorry if I'm coming across as pushy, not meaning to be at all. Just throwing my thoughts down as and when they come to me :) 

:+1:

**Comment by drvinceknight** [comment]
> I see that in your current version you're still looping the plays of the Matches)

I know this is all temporary: sorry, I realise that could have come across as pushy. Sorry! Just my usual blurting out what I'm thinking :) :+1: Sorry :)

**Comment by marcharper** [comment]
It's totally fine, we're thinking along the same lines so it's all good. I've literally never touched this part of the library so it's good to have you also thinking through the various concerns. But you are getting ahead of me, and we're duplicating efforts at this point.

**Comment by drvinceknight** [comment]
> But you are getting ahead of me, and we're duplicating efforts at this point.

Yup. Shutting up now, will be patient (just excited about this). :) :+1:

**Comment by marcharper** [comment]
See [this gist](https://gist.github.com/marcharper/516cbc2061c0556d8c341008009dbfbe) and the comment there. Writing to disk incurs almost no difference in performance, and total memory is greatly reduced (and runtime seems better as well). Comments are welcome, if there are objections that would be good to know before I try to integrate more deeply into the library.

**Comment by drvinceknight** [comment]
No objections from me: I think this looks fantastic. Slightly raising my eyebrow at the `time.sleep(0.01)` but I'm not sure there's anything to be done and indeed if it's that bad. :+1: 

(FYI. I prefer your more generic `generate_match_parameters` here than the one I'd seen before where you split out the players based on classification. I think this is nicer.)

**Comment by drvinceknight** [comment]
Hey @marcharper. With #565 the order of play of matches is not important from the point of view of results and data files (I'm not talking about the cache here) so why would the following not work (you might have explained this already,  apologies, just thinking about the time.sleep) :

- modify the current generator class to incorporate the repetitions and yield 'chunks' of matches/match parameters (the chunk size could be a parameter). 
- use the current setup of workers etc to work through those chunks. 
- similar to the piecewise writing fix PR I closed (the one you tried and said helped a bit), we append/write to file the results of the chunks as we go

I'm probably missing something but I believe that would be a simple enough modification that would get rid of the problem of handling huger numbers of interactions. 

I might throw that together later to see how that works... 

I'm guessing I'm missing something about threads and other voodoo like that.

**Comment by drvinceknight** [comment]
https://github.com/Axelrod-Python/Axelrod/compare/matchchunks shows that. I haven't fixed all the tests there and there is the bigger thing that needs fixing which is that I'm not actually using the generator and turning it in to a list immediately for the parallel processing (but that's fixable, I guess...). I've run some tournaments locally and they seem to behave.

I'm still guessing I'm missing something.

**Comment by marcharper** [comment]
There are a number of issues including:
* using the cache effectively
* memory usage per process
* sharing state properly between processes
* being able to follow the logic of the code

When python forks a thread it allocates the same amount of memory as the calling process. That causes a number of issues, and it's best not to hold results in the calling thread, not to create a bunch of matches that are not yet used, etc. (It would be easier / more simple if we always wrote the results to disk.) The other benefit to chunking is that the cache has fewer entries and the hash table lookups are much faster.

There's nothing wrong with using `time.sleep` -- it's just to limit cpu on the thread that's waiting on results to consume. I'm not claiming that my solution is the best possible, but so far it appears to be:
* fast
* uses memory much more efficiently
* is more generic / flexible
* simplifies a lot of existing code in the library

I've been running tests for the past few days while I'm busy with other things, and have generated 20+ gb of match data with my code without taking down the machine (whereas the older code couldn't get through a tournament of 100 repetitions). There are a lot of changes that need to be tested. I ask for patience in the meantime, and feel free to tinker with the parallel branch, which I'm trying to keep updated as I go. At the moment it contains the following changes:
* addition of the new parallel processing functions
* rewrite of the the match generators
* rewrite of the tournament code to use the new parallel processing functions
* removal of a lot of now unnecessary code

I believe there is just one issue left, getting the data back out of the consumer process in the case that it isn't written to disk. This is why the tests are still failing. Python doesn't make this easy and I may resort to a tempfile until I figure out a better solution.

**Comment by drvinceknight** [comment]
> using the cache effectively
> memory usage per process
> sharing state properly between processes
> being able to follow the logic of the code

> When python forks a thread it allocates the same amount of memory as the calling process. That causes a number of issues, and it's best not to hold results in the calling thread, not to create a bunch of matches that are not yet used, etc. (It would be easier / more simple if we always wrote the results to disk.) The other benefit to chunking is that the cache has fewer entries and the hash table lookups are much faster.

Cool. I still don't quite understand what my approach (which uses chunking by the way) doesn't do, (the cache could be plugged back in to deal with the repetitions, the chunk size deals with memory etc, the matches could be replaced by parameters etc... my branch was just a quick concept of using chunking to replace the repetitions) but don't worry I'm sure there is a lot I don't understand about threading and stuff.

> I ask for patience in the meantime, and feel free to tinker with the parallel branch, which I'm trying to keep updated as I go. 

I am not being impatient. I just had a thought, of how I thought it would be straightforward to include chunking that I thought could be helpful as you work through this but I can just wait till you've done it all. 

I was also asking why my approach wouldn't work or what it doesn't do because I'm curious (and believe I don't understand something) but perhaps that will become clear through the PR.

> addition of the new parallel processing functions
> rewrite of the the match generators
> rewrite of the tournament code to use the new parallel processing functions
> removal of a lot of now unnecessary code

I noticed you got rid of the match generators which seemed strange as I felt they're quite a nice way to make new tournaments but I thought probably best wait to see how it all fits together (am being patient). We can just talk about that on the PR.

> I believe there is just one issue left, getting the data back out of the consumer process in the case that it isn't written to disk. This is why the tests are still failing. Python doesn't make this easy and I may resort to a tempfile until I figure out a better solution.

I think this shows that I'm missing quite a lot about this (threading etc...) Which I mentioned when I asked why my idea wouldn't work. Not sure if you saw what I had on my branch but it seems to be able to get that back easily enough. No doubt I'm not quite understanding threads and stuff (I really don't doubt that: it's all voodoo to me :)).

👍

**Comment by marcharper** [comment]
There is a new match generator function in the parallel processing code that generates matches for both types of tournaments. It was easier to just write a single function to do so and there's no need to have classes with inheritance to do so.

I didn't run your code but I suspect that it works fine for small tournaments and has the same issues as the previous code for larger tournaments, tournaments with more turns, and any other variation to causes memory accumulation. I actually tried something very similar at first, trying to change the library as little as possible, but it didn't work out well for some of the usage cases I tested. (That doesn't mean it can't work of course but I didn't figure out quite how if it is possible.) However if you want to run some of those cases and they work out then I'm happy to consider an alternate approach.

**Comment by drvinceknight** [comment]
> There is a new match generator function in the parallel processing code that generates matches for both types of tournaments. It was easier to just write a single function to do so and there's no need to have classes with inheritance to do so.

Ok. Feels like that should be in a different module (as hopefully they will be used even in non parallel processing cases) but again it's probably best to wait till you're all done so sounds good. Simplifying to a function from a class sounds good: I'm guessing the idea being that people could write new functions if they need to (as opposed to inherit classes). 👍 

> 
> I didn't run your code but I suspect that it works fine for small tournaments and has the same issues as the previous code for larger tournaments, tournaments with more turns, and any other variation to causes memory accumulation. I actually tried something very similar at first, trying to change the library as little as possible, but it didn't work out well for some of the usage cases I tested. (That doesn't mean it can't work of course but I didn't figure out quite how if it is possible.) However if you want to run some of those cases and they work out then I'm happy to consider an alternate approach.

Wasn't expecting you to run it. Just hoping you might take a look at how I implemented the chunking and let me know what I'm missing but I know you're busy and it's not your responsibility to teach me about parallel processing. 

I'll wait for the PR, good luck figuring out that last part about getting the data out when not writing to disk :) 👍

**Comment by marcharper** [comment]
It's certainly good to have multiple approaches. The best thing to do would be to run them both and look at the max memory usage per process and total runtime for a few different usage cases.

We can reorganize the parallel code if we go that route, I'm fine with moving the generators around, and for the moment I just want all the tests to pass. 

I should have some time tonight to look at it.

The challenge with the chunking is that there is a huge qualitative difference in the performance/memory usage in at least three cases:
* both strategies deterministic, very fast with the cache and a lot of data is generated quickly
* some strategies store a lot more memory (metastrategies?) and python doesn't seem to release the objects until all the matches are done
* some strategies are much more cpu intensive (e.g. cycle hunter) and take a lot of time but not much memory

So it's tricky to get the chunking right to reduce both memory accumulation and reduce overhead. There's also the number of repetitions to deal with...

**Comment by drvinceknight** [comment]
As you mentioned, runtime and memory is not everything. I'd be worried
about over optimising the chunking so that it's very specific to the bigger
tournament and potentially not clear code (_I haven't seen how you've got
this at present so don't know if that's the case_).

I understand those complications you mentioned. If you get a chance to
actually look at what I've written (_which as I said is still an early
concept_) that'd be great but not a problem if not. I'll keep going with it
at some point so we can then have an actual discussion and figure out
what's best. :) :+1:

(FYI Either way replacing the generator classes with functions makes sense
to me. At some point I thought perhaps having them as tournament methods
could be worth doing, not sure...)

**Comment by drvinceknight** [comment]
To be sure I'm clear (this is what I tried to say in my first comment this morning): 

- **I VERY much doubt my approach is going to be better than yours**, you have looked at this far longer than me and know these things far better than I do. 
- I was/am just asking why the simpler approach doesn't do the trick (as far as I can see it's aiming to do the same overall thing). Perhaps once it's 'finished' I'll understand.

**Comment by marcharper** [comment]
@drvinceknight @meatballs 

Here's what I've learned lately. In the current master branch, the memory of each worker thread is basically bounded below by the memory of the calling thread, which is ~400MB. I believe this is due to building / loading the cache and storing those interactions before forking, so it's dependent on `turns` and number of players squared (but not repetitions). It's bounded above roughly by `repetitions` * `turns` * matches per repetition (~ players * players / 2 ).

We can deal with the memory usage in a few ways:
* simply modify Tournament to write each repetition's interactions to disk. We're still sunk by e.g. turns == 1000 or a large number of players on each thread, but ok on repetitions, since we're not holding everything in tournament
* modify Tournament to write individual match results to its queue rather than a full repetition's results, and write this queue to disk as the workers run. I believe this is essentially what @drvinceknight proposed above (sans chunking). This uses memory approximately equal to (base thread memory) * (number of workers). We're still hit by the base thread memory but approx. a factor of 2 better per thread than the previous approach
* further limit memory by not building a full cache or saving any interactions in the base thread, letting the matches generate individual caches as needed (e.g. match chunking). This gets us to typically (50-100 MB) * (number of workers). This is what I've done on the parallel branch but this may be more simply achieved by taking the previous approach a built further -- modifying the order in which players are generated and how the cache is used (generated by each match rather than for all players).

Another benefit of the last approach is a performance gain on cache lookups by computing along matches then repetitions (rather than repetitons then matches). This increases with the number of repetitions. The total cpu difference is small -- ~7% total cpu time -- for 20 repetitions. For many more repetitions it's hard to say because the current approach takes too much memory (and eventually hits the swap on my box, which kill performance).

Regardless of what we do above, we can't continue holding all the interactions in memory. It's also an issue for building a result_set from disk, since a standard tournament of 20 repetitions, 200 turns, and 113 strategies takes 2 GB of memory to load and right now we load all the interactions from disk. (This scales linearly in repetitions and turns, and quadratically in strategies.)

My recommendation at this time is the last approach described above in a significantly less complicated way then I've done on the parallel branch. My branch makes a lot of changes to the library, only some of which are ultimately necessary for the benefits described above, and the benefits beyond what I've described above may not be worth it -- e.g. being able to use the standard multiprocessing library at the expense of more complicated match generators, and various changes to accommodate. I recommend that we
* move most of the tournament computing functionality out to a separate class that takes a generic match generator and handles all the match computing in a generic way (including any caching). This will be good for tournaments and also anyone that wants to generate a bunch of matches using parallel processing.
* rewrite the match generators in chunks (by repetitions) to take maximum advantage of the cache
* write interactions to disk by default rather than keep interactions in memory
* modify resultset to be able to handle larger data sets (computing the necessary quantities without keeping all the interactions in memory, if possible)
* We should further change the results data file to be:
index1, index2, name1, name2, history1, history2

If everyone's happy with this, I'll make a new branch with these changes, using what I've done on the parallel branch as appropriate, along with the commits that @drvinceknight made earlier for my parallel branch.

**Comment by marcharper** [comment]
The speed up on my parallel branch does seem to improve with `repetitions`. On master a 200 turn tournament of 113 players and 20 repetitions takes 36 minutes (CPU time) versus 33 on my branch. However for 100 repetitions it takes 148 minutes (rather than the 180 I would have expected for master).

**Comment by drvinceknight** [comment]
Huge piece of work @marcharper! I know you've been working loads on this and collecting a bunch of insights about run times and stuff. This is awesome :)

I think my vote would be for your third suggestion.

> move most of the tournament computing functionality out to a separate class that takes a generic match generator and handles all the match computing in a generic way (including any caching). This will be good for tournaments and also anyone that wants to generate a bunch of matches using parallel processing.

Making sure I understand: this is basically creating an abstract step of a way to run *any collection of matches**, of which a tournament is one? What would the role of the tournament be? Mainly user facing and pulling everything together? So a tournament would have two main engines: "a match generator" and "a match player"?

Sounds good to me. 

> rewrite the match generators in chunks (by repetitions) to take maximum advantage of the cache

Yup, sounds good to me.

I have a "thought" at the end of this comment about this.

> write interactions to disk by default rather than keep interactions in memory 

I think I would suggest/prefer we keep things in memory when running things serially. These would usually be smaller tournaments that would run ok? (Thinking amongst other things of the very beginner level user who might be confused by csv files appearing.)

When running in series we stay in memory by default and use temp files in parallel by default: so that we keep the `results = tournament.play()` functionality which is backwards compatible. 

> (alternatively we can do a calculation at runtime to see if keeping the results in memory is ok, or warn the user somehow)

My "thought" coming at the end might have a way to have a warning.

> modify resultset to be able to handle larger data sets (computing the necessary quantities without keeping all the interactions in memory, if possible)

Great idea. I think this could be another issue/PR? 

Perhaps (thinking about readability), we could have a `BigResultsSet` which does this? 

Mainly thinking of the current result set which I believe is quite readable (I might very well be wrong here and no one would hurt my feelings if we felt it wasn't). Of course it might be possible to keep the readability and also get your suggest working. Either way I suggest it's a second issue/PR for now.

> We should further change the results data file to be: index1, index2, name1, name2, history1, history2 This is generically more useful IMO.

Absolutely. This is way better. :+1: 

(As a place marker: there's currently a utility function that takes the current string rep and outputs interactions, it should be changed to complement this.)

> The speed up on my parallel branch does seem to improve with repetitions. On master a 200 turn tournament of 113 players and 20 repetitions takes 36 minutes (CPU time) versus 33 on my branch. However for 100 repetitions it takes 148 minutes on the parallel branch (rather than the 180 I would have expected for master).

The speed up is really great **but** it's mainly about reducing the memory which I think you've basically smashed here :) :+1: 

---

My thought (which could just be a discussion on the PR - I will not insist about this here):

This is about re writing the generators (which should be done and 100% on board with chunking).  I know we've discussed using a function vs a class.

I like the idea of simplifying things to the level of a function but **IF** (big if) we want to have a progress bar (which is not at all important, so this really is a side point but could perhaps indicate other more important functionality we may want/need) then having a class would be easier. 

This is because to have a progress bar we need to know the total number of matches (or chunks, just depends on granularity of the progress bar). As we are using generators (which we have to for memory) we can't know the number of matches until they're all done.

For some/most match generators it's easy enough to calculate the number of matches we will have (for a round robin it's some binomial coefficient involving the number of players etc). So I suggest we have a `__len__` attribute in the generator:

```
     self.__len__(self):
           return self.repetitions * self.nplayers  # or whatever the actual formula is here
```

Then we could have tests that check this:

```
g = RoundRobinGenerator()
matches = list(g.generate()) # Probably needing to flatten out the chunks etc, hopefully you get the idea
assert(len(matches), len(g))
```

Some generators **will not be able to have a length**, for example the ones I make in the tutorial on writing new tournaments: probabilistically deciding to have a match between two given players. We would then just have:

```
     self.__len__(self, ....):
           return None # Or raise a warning or something
```

Then for the progress bar, we could just raise a warning saying that "For this collection of matches, no progress bar is available".

**Whilst the above is mainly about a progress bar** I think we could have similar things that could display a warning as you suggested for 'larger' collections of matches, so in the `init`, we could have a `large` attribute set to `False` or `True` based on the size etc and then raise a warning if need be. "Hey you should think of writing this to disk because it's gynormous". I can imagine there might be further characteristics about a collection of matches that we might want attributes for (all deterministic? etc...).

**Comment by marcharper** [comment]
Yep, that's what I'm planning for tournaments and the "MatchPlayer" class that handles the parallel aspects.

I actually think a class is better for match generating now, and I'm on board with your suggestions for progress bars and potentially other methods. If we decide to move to a function instead of a class we can do it in a later PR. What I've done on the parallel branch is ultimately not worth the complexity and loss of readability IMO.

We can write to disk using tempfiles so that the user won't notice any extraneous CSV files. I think in the case that the match generator doesn't report a length (or can't) that disk by default is the way to go. But we can work out the details on that once the other changes are in.

**Comment by drvinceknight** [comment]
All sounds great to me. :+1: 

Happy to help out if I can, just let me know.

**Comment by marcharper** [comment]
I'm working on the parallel branch, could use your help with the tests @drvinceknight .

**Comment by drvinceknight** [comment]
Cool, I'll take a look later today will send you a PR.

**Comment by drvinceknight** [comment]
Closed by #572
